File: src/bin/lc61.cpp
#include <iostream>
struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

ListNode* rotateRight(ListNode* head, int k) {
  if (head == nullptr) {
    return nullptr;
  }
  auto p = head;
  int len = 1;
  while (p->next != nullptr) {
    p = p->next;
    len++;
    /* code */
  }
  auto step = k % len - 1;
  p->next = head;
  p = head;
  //   int n = k - ;
  while (step > 0) {
    step--;
    p = p->next;
  }
  auto rst = p->next;
  p->next = nullptr;
  return rst;
}

void dbg_print(ListNode* p) {
  using namespace std;
  while (p != nullptr) {
    /* code */
  }
}

File: src/bin/lc61.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

fn main() {}

pub fn rotate_right(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
    // {
    let mut head = head;
    let mut p = head.as_mut();
    {
        while let Some(node) = p {
            if node.next.as_ref().is_some() {
                p = node.next.as_mut();
            } else {
                node.next.as_mut().unwrap().next = head.take();
            }
        }
    }
    // }
    head
}
